readme.md
<p align="center"><a href="#installation">Installation</a> • <a href="#first-steps">First steps</a> • <a href="#rule-examples">Rule examples</a> • <a href="#viewer">Viewer</a> • <a href="#usage">Usage</a> • <a href="#build-status">Build Status</a> • <a href="#license">License</a></p>

<p align="center">
<img width="300" height="150" src="https://essentialkaos.com/github/mockka.png"/>
</p>

`Mockka` is a utility for mocking and testing HTTP API's.

## Installation
To build the Mockka from scratch, make sure you have a working Go 1.5+ workspace ([instructions](https://golang.org/doc/install)), then:

```
go get github.com/essentialkaos/mockka
```

If you want update Mockka to latest stable release, do:

```
go get -u github.com/essentialkaos/mockka
```

## First steps

Show basic info about all mock files:
````
mockka -c /path/to/mockka.conf list
````

Run Mockka server:
````
mockka -c /path/to/mockka.conf run
````

By default Mockka try to find configuration file in next locations:

* `/etc/mockka.conf`
* `~/.mockka.conf` (`$HOME` directory)
* `./mockka.conf` (current directory)

## Rule examples

Latest version of Mockka have full [Fake](https://github.com/icrowley/fake) package support. List of all supported functions can be found [here](https://github.com/essentialkaos/mockka/wiki/Supported-Stubber-Methods).

#### Example 1
````bash
# This is comment

# Description is optional field
@DESCRIPTION
Example mock file #1

# Basic auth login and password
@AUTH
user:password

@REQUEST
GET /test/login?action=login&type=1

# You can use go template language for response body
@RESPONSE
{{ if .HeaderIs "X-Suppa-Header" "Magic" }}
{
  "status": "ok",
  "result": {
    "username":"{{ .UserName "en" }}",
    "ip": "{{ .IPv4 }}",
    "first_name": "{{ .MaleFirstName "ru" }}",
    "last_name": "{{ .MaleLastName "ru" }}",
    "city": "{{ .City "ru" }}",
    "password": "{{ .SimplePassword }}",
    "action": "{{ .Query "action" }}"
  }
}
{{ else }}
{
  "status": "error",
  "result": {
    "username":"{{ .UserName "en" }}",
    "action": "{{ .Query "action" }}"
  }
}
{{ end }}

# Code is optional field, 200 by default
@CODE
200

@HEADERS
Content-Type:application/json
X-Content-Type-Options:nosniff
X-Seraph-LoginReason:OK

# Delay has float type
@DELAY
2.0

````

#### Example 2

````bash
@DESCRIPTION
Example mock file #2

@REQUEST
GET /test/login?action=login&type=1

# You can define different responses in one mock file
@RESPONSE:1
{
  "status": "ok",
  "result": {
    "username":"{{ .UserName "en" }}"
  }
}

@RESPONSE:2
{
  "status": "error",
  "result": {}
}

# You can define different http codes for each response
@CODE:1
200

@CODE:2
502

# Or define default value for all responses
@HEADERS
Content-Type:application/json
X-Content-Type-Options:nosniff
X-Seraph-LoginReason:OK

@DELAY:1
2.0

@DELAY:2
15

````

#### Example 3

````bash
@DESCRIPTION
Example mock file #3

# If you use hostnames instead ip you can define it here
@HOST
my-suppa-host.domain.com

@REQUEST
GET /test/login?action=login&type=1

# You can store each response in file
@RESPONSE:1 < example-error.json
@RESPONSE:2 < example-ok.json
@RESPONSE:3 < example-unknown-user.json

# Or return body from proxied request
@RESPONSE:4 < https://api.domain.com/someurl

# Also you can overwrite headers and status code from rule by
# values from proxied request response
@RESPONSE:5 << https://api.domain.com/someurl

@CODE
200

@HEADERS
Content-Type:application/json

````

#### Example 4 (wildcard query)

````bash
@DESCRIPTION
Example mock file #4

@REQUEST
GET /test/status?username=*&type=1

# You can define query params in response body
@RESPONSE
{
  "status": "ok",
  "result": {
    "username": "{{ .Query "username" }}",
  }
}

@HEADERS
Content-Type:application/json

````

## Viewer

For viewing mockka logs we provide simple tool named `mockka-viewer`.

Some features:

* Syntax higlighting
* Filtering records by time range
* Log file search

Usage:

````
Usage: mockka-viewer <options> log-file

Options:

  --no-color, -nc     Disable colors in output
  --from, -f date     Time range start
  --to, -t date       Time range end
  --help, -h          Show this help message
  --version, -v       Show version

Examples:

  mockka-viewer /path/to/file.log
  Read log file

  mockka-viewer file.log
  Try to find file.log in mockka logs directory

  mockka-viewer file
  Try to find file.log in mockka logs directory

  mockka-viewer -f 2016/01/02 -t 2016/01/05 /path/to/file.log
  Read file and show only records between 2016/01/02 and 2016/01/05

  mockka-viewer -f "2016/01/02 12:00" /path/to/file.log
  Read file and show only records between 2016/01/02 12:00 and current moment

````

## Usage

```
Usage: mockka <command> <options>

Commands:

  run                  Run mockka server
  check mock-file      Check rule for problems
  make mock-name       Create mock file from template
  list service-name    Show list of exist rules

Options:

  --config, -c file        Path to config file
  --port, -p 1024-65535    Overwrite port
  --daemon, -d             Run server in daemon mode
  --no-color, -nc          Disable colors in output
  --help, -h               Show this help message
  --version, -v            Show version

Examples:

  mockka -c /path/to/mockka.conf run
  Run mockka server and use config /path/to/mockka.conf

  mockka make service1/test1
  Create file test1.mock for service service1.

  mockka check service1/test1
  Check rule file test1.mock for service service1

  mockka check service1/test1
  Check all rules of service service1

  mockka list
  List all rules

  mockka list service1
  List service1 rules

```

## Build Status

| Branch | Status |
|------------|--------|
| `master` | [![Build Status](https://travis-ci.org/essentialkaos/mockka.svg?branch=master)](https://travis-ci.org/essentialkaos/mockka) |
| `develop` | [![Build Status](https://travis-ci.org/essentialkaos/mockka.svg?branch=develop)](https://travis-ci.org/essentialkaos/mockka) |

## License

[EKOL](https://essentialkaos.com/ekol)


changelog.md
### Changelog

#### 1.7.4

* [EK](https://github.com/essentialkaos/ek) package updated to v3

#### 1.7.3

* Improved debug output
* Fixed bug with global mock files placed in rules directory
* Fixed bug with rendering templates

#### 1.7.2

* Improved debug output
* Increased coverage
* Improved compatibility with latest version of ek package

#### 1.7.0

* Mock files validation feature (`mockka check ...`)
* Log file viewer (`mockka-viewer`)
* Request proxying feature (you can proxy request to another URI and use result as response body)
* Config searching feature (we search valid config in /etc, home directory and current directory)
* Auto head feature (mockka automatically generate response for HEAD request for any another requests)
* [Fake](https://github.com/icrowley/fake) package support in stabber
* Added two types of logging requests - separate and united
* Fully rewrited working with wildcard request
* Improved rule directory scan
* Many new improvements from latest version of [EK](https://github.com/essentialkaos/ek) package
* Improved logging
* Fixed some minor bugs


cli/cli.go
package cli

// ////////////////////////////////////////////////////////////////////////////////// //
//                                                                                    //
//                     Copyright (c) 2009-2016 Essential Kaos                         //
//      Essential Kaos Open Source License <http://essentialkaos.com/ekol?en>         //
//                                                                                    //
// ////////////////////////////////////////////////////////////////////////////////// //

import (
	"fmt"
	"os"
	"runtime"

	"pkg.re/essentialkaos/ek.v3/arg"
	"pkg.re/essentialkaos/ek.v3/fmtc"
	"pkg.re/essentialkaos/ek.v3/fsutil"
	"pkg.re/essentialkaos/ek.v3/knf"
	"pkg.re/essentialkaos/ek.v3/log"
	"pkg.re/essentialkaos/ek.v3/signal"
	"pkg.re/essentialkaos/ek.v3/system"
	"pkg.re/essentialkaos/ek.v3/usage"

	"github.com/essentialkaos/mockka/generator"
	"github.com/essentialkaos/mockka/listing"
	"github.com/essentialkaos/mockka/rules"
	"github.com/essentialkaos/mockka/server"
	"github.com/essentialkaos/mockka/validator"
)

// ////////////////////////////////////////////////////////////////////////////////// //

const (
	APP  = "Mockka"
	VER  = "1.7.4"
	DESC = "Utility for mockking HTTP API's"
)

const (
	ARG_CONFIG   = "c:config"
	ARG_PORT     = "p:port"
	ARG_DAEMON   = "d:daemon"
	ARG_NO_COLOR = "nc:no-color"
	ARG_HELP     = "h:help"
	ARG_VER      = "v:version"
)

const (
	MIN_PORT          = 1024
	MAX_PORT          = 65535
	MIN_READ_TIMEOUT  = 1
	MAX_READ_TIMEOUT  = 120
	MIN_WRITE_TIMEOUT = 1
	MAX_WRITE_TIMEOUT = 120
	MIN_HEADER_SIZE   = 1024
	MAX_HEADER_SIZE   = 10 * 1024 * 1024
	MIN_CHECK_DELAY   = 1
	MAX_CHECK_DELAY   = 3600
)

const (
	MAIN_DIR                  = "main:dir"
	DATA_RULE_DIR             = "data:rule-dir"
	DATA_LOG_DIR              = "data:log-dir"
	DATA_LOG_TYPE             = "data:log-type"
	DATA_CHECK_DELAY          = "data:check-delay"
	HTTP_IP                   = "http:ip"
	HTTP_PORT                 = "http:port"
	HTTP_READ_TIMEOUT         = "http:read-timeout"
	HTTP_WRITE_TIMEOUT        = "http:write-timeout"
	HTTP_MAX_HEADER_SIZE      = "http:max-header-size"
	HTTP_MAX_DELAY            = "http:max-delay"
	PROCESSING_AUTO_HEAD      = "processing:auto-head"
	PROCESSING_ALLOW_PROXYING = "processing:allow-proxying"
	LOG_DIR                   = "log:dir"
	LOG_FILE                  = "log:file"
	LOG_PERMS                 = "log:perms"
	LOG_LEVEL                 = "log:level"
	ACCESS_USER               = "access:user"
	ACCESS_GROUP              = "access:group"
	ACCESS_MOCK_PERMS         = "access:mock-perms"
	ACCESS_LOG_PERMS          = "access:log-perms"
	ACCESS_MOCK_DIR_PERMS     = "access:mock-dir-perms"
	ACCESS_LOG_DIR_PERMS      = "access:log-dir-perms"
	LISTING_SCHEME            = "listing:scheme"
	LISTING_HOST              = "listing:host"
	LISTING_PORT              = "listing:port"
	TEMPLATE_PATH             = "template:path"
)

const (
	COMMAND_RUN   = "run"
	COMMAND_LIST  = "list"
	COMMAND_MAKE  = "make"
	COMMAND_CHECK = "check"
)

// ////////////////////////////////////////////////////////////////////////////////// //

var argMap = arg.Map{
	ARG_CONFIG:   &arg.V{Value: "/etc/mockka.conf"},
	ARG_PORT:     &arg.V{Type: arg.BOOL, Min: MIN_PORT, Max: MAX_PORT},
	ARG_DAEMON:   &arg.V{Type: arg.BOOL},
	ARG_NO_COLOR: &arg.V{Type: arg.BOOL},
	ARG_HELP:     &arg.V{Type: arg.BOOL, Alias: "u:usage"},
	ARG_VER:      &arg.V{Type: arg.BOOL, Alias: "ver"},
}

// ////////////////////////////////////////////////////////////////////////////////// //

func Init() {
	var err error
	var errs []error

	runtime.GOMAXPROCS(1)

	if len(os.Args) <= 1 {
		showUsage()
		return
	}

	args, errs := arg.Parse(argMap)

	if len(errs) != 0 {
		for _, err := range errs {
			fmt.Println(err.Error())
		}

		os.Exit(1)
	}

	if arg.GetB(ARG_NO_COLOR) {
		fmtc.DisableColors = true
	}

	if arg.GetB(ARG_VER) {
		showAbout()
		return
	}

	if len(args) == 0 || arg.GetB(ARG_HELP) {
		showUsage()
		return
	}

	// Slice with valid config paths
	confPaths := []string{
		arg.GetS(ARG_CONFIG),
		"/etc/mockka.conf",
		"~/mockka.conf",
		"mockka.conf",
	}

	err = knf.Global(fsutil.ProperPath("FRS", confPaths))

	if err != nil {
		fmt.Println(err.Error())
		os.Exit(1)
	}

	errs = validateConfig()

	if len(errs) != 0 {
		for _, err := range errs {
			fmt.Println(err.Error())
		}

		os.Exit(1)
	}

	setupLog()

	if arg.GetB(ARG_DAEMON) {
		registerSignalHandlers()
	}

	execCommand(args)
}

func validateConfig() []error {
	var permsChecker = func(config *knf.Config, prop string, value interface{}) error {
		if !fsutil.CheckPerms(value.(string), config.GetS(prop)) {
			switch value.(string) {
			case "DRX":
				return fmt.Errorf("Property %s must be path to readable directory.", prop)
			}
		}

		return nil
	}

	var userChecker = func(config *knf.Config, prop string, value interface{}) error {
		if config.GetS(prop) == "" {
			return nil
		}

		if !system.IsUserExist(config.GetS(prop)) {
			return fmt.Errorf("User defined in %s is not exist.", prop)
		}

		return nil
	}

	var groupChecker = func(config *knf.Config, prop string, value interface{}) error {
		if config.GetS(prop) == "" {
			return nil
		}

		if !system.IsGroupExist(config.GetS(prop)) {
			return fmt.Errorf("Group defined in %s is not exist.", prop)
		}

		return nil
	}

	return knf.Validate([]*knf.Validator{
		&knf.Validator{DATA_RULE_DIR, knf.Empty, nil},
		&knf.Validator{DATA_LOG_DIR, knf.Empty, nil},
		&knf.Validator{DATA_CHECK_DELAY, knf.Empty, nil},
		&knf.Validator{HTTP_PORT, knf.Empty, nil},
		&knf.Validator{HTTP_READ_TIMEOUT, knf.Empty, nil},
		&knf.Validator{HTTP_WRITE_TIMEOUT, knf.Empty, nil},
		&knf.Validator{HTTP_MAX_HEADER_SIZE, knf.Empty, nil},
		&knf.Validator{ACCESS_MOCK_PERMS, knf.Empty, nil},
		&knf.Validator{ACCESS_LOG_PERMS, knf.Empty, nil},
		&knf.Validator{ACCESS_MOCK_DIR_PERMS, knf.Empty, nil},
		&knf.Validator{ACCESS_LOG_DIR_PERMS, knf.Empty, nil},

		&knf.Validator{DATA_RULE_DIR, permsChecker, "DRX"},
		&knf.Validator{DATA_LOG_DIR, permsChecker, "DRX"},

		&knf.Validator{DATA_CHECK_DELAY, knf.Less, MIN_CHECK_DELAY},
		&knf.Validator{DATA_CHECK_DELAY, knf.Greater, MAX_CHECK_DELAY},
		&knf.Validator{HTTP_PORT, knf.Less, MIN_PORT},
		&knf.Validator{HTTP_PORT, knf.Greater, MAX_PORT},
		&knf.Validator{HTTP_READ_TIMEOUT, knf.Less, MIN_READ_TIMEOUT},
		&knf.Validator{HTTP_READ_TIMEOUT, knf.Greater, MAX_READ_TIMEOUT},
		&knf.Validator{HTTP_WRITE_TIMEOUT, knf.Less, MIN_WRITE_TIMEOUT},
		&knf.Validator{HTTP_WRITE_TIMEOUT, knf.Greater, MAX_WRITE_TIMEOUT},
		&knf.Validator{HTTP_MAX_HEADER_SIZE, knf.Less, MIN_HEADER_SIZE},
		&knf.Validator{HTTP_MAX_HEADER_SIZE, knf.Greater, MAX_HEADER_SIZE},

		&knf.Validator{ACCESS_USER, userChecker, nil},
		&knf.Validator{ACCESS_GROUP, groupChecker, nil},
	})
}

func setupLog() {
	var err error

	if arg.GetB(ARG_DAEMON) {
		err = log.Set(knf.GetS(LOG_FILE), knf.GetM(LOG_PERMS, 0644))

		if err != nil {
			fmt.Printf("Can't setup logger: %v\n", err)
			os.Exit(1)
		}
	}

	err = log.MinLevel(knf.GetS(LOG_LEVEL, "info"))

	if err != nil {
		fmt.Printf("Can't set log level: %v\n", err)
	}
}

func execCommand(args []string) {
	command := args[0]

	switch command {
	case COMMAND_RUN:
		runServer()

	case COMMAND_MAKE:
		makeMock(args[1:])

	case COMMAND_LIST:
		listMocks(args[1:])

	case COMMAND_CHECK:
		checkMocks(args[1:])

	default:
		printError(fmt.Sprintf("Unknown command %s", command))
		os.Exit(1)
	}
}

func registerSignalHandlers() {
	signal.Handlers{
		signal.INT:  intSignalHandler,
		signal.TERM: termSignalHandler,
		signal.HUP:  hupSignalHandler,
	}.TrackAsync()
}

func runServer() {
	observer := rules.NewObserver(knf.GetS(DATA_RULE_DIR))
	observer.AutoHead = knf.GetB(PROCESSING_AUTO_HEAD)
	observer.Start(knf.GetI(DATA_CHECK_DELAY))

	err := server.Start(observer, APP+"/"+VER, arg.GetS(ARG_PORT))

	if err != nil {
		printError(err.Error())
		os.Exit(1)
	}
}

func makeMock(args []string) {
	var name = ""

	if len(args) != 0 {
		name = args[0]
	}

	err := generator.Make(name)

	if err != nil {
		printError(err.Error())
		os.Exit(1)
	}
}

func listMocks(args []string) {
	var service = ""

	// Suppress observer logging
	log.Set(os.DevNull, 0)

	if len(args) != 0 {
		service = args[0]
	}

	observer := rules.NewObserver(knf.GetS(DATA_RULE_DIR))
	err := listing.List(observer, service)

	if err != nil {
		printError(err.Error())
		os.Exit(1)
	}
}

func checkMocks(args []string) {
	var mock = ""

	if len(args) != 0 {
		mock = args[0]
	}

	err := validator.Check(mock)

	if err != nil {
		printError(err.Error())
		os.Exit(1)
	}
}

func printError(message string) {
	if arg.GetB(ARG_DAEMON) {
		fmt.Printf("\n%s\n\n", message)
	} else {
		fmtc.Printf("\n{r}%s{!}\n\n", message)
	}
}

// ////////////////////////////////////////////////////////////////////////////////// //

func intSignalHandler() {
	log.Info("Received INT signal, shutdown...")
	os.Exit(0)
}

func termSignalHandler() {
	log.Info("Received TERM signal, shutdown...")
	os.Exit(0)
}

func hupSignalHandler() {
	log.Info("Received HUP signal, log reopened")
	log.Reopen()
}

// ////////////////////////////////////////////////////////////////////////////////// //

func showUsage() {
	info := usage.NewInfo("")

	info.AddCommand(COMMAND_RUN, "Run mockka server")
	info.AddCommand(COMMAND_CHECK, "Check rule for problems", "mock-file")
	info.AddCommand(COMMAND_MAKE, "Create mock file from template", "mock-name")
	info.AddCommand(COMMAND_LIST, "Show list of exist rules", "service-name")

	info.AddOption(ARG_CONFIG, "Path to config file", "file")
	info.AddOption(ARG_PORT, "Overwrite port", fmt.Sprintf("%d-%d", MIN_PORT, MAX_PORT))
	info.AddOption(ARG_DAEMON, "Run server in daemon mode")
	info.AddOption(ARG_NO_COLOR, "Disable colors in output")
	info.AddOption(ARG_HELP, "Show this help message")
	info.AddOption(ARG_VER, "Show version")

	info.AddExample(
		"-c /path/to/mockka.conf run",
		"Run mockka server and use config /path/to/mockka.conf",
	)

	info.AddExample(
		"make service1/test1",
		"Create file test1.mock for service service1.",
	)

	info.AddExample(
		"check service1/test1",
		"Check rule file test1.mock for service service1",
	)

	info.AddExample(
		"check service1/test1",
		"Check all rules of service service1",
	)

	info.AddExample("list", "List all rules")
	info.AddExample("list service1", "List service1 rules")

	info.Render()
}

func showAbout() {
	about := &usage.About{
		App:     APP,
		Version: VER,
		Desc:    DESC,
		Year:    2009,
		Owner:   "ESSENTIAL KAOS",
		License: "Essential Kaos Open Source License <https://essentialkaos.com/ekol?en>",
	}

	about.Render()
}
